non-normative/recipes.adoc

```diff
@@ -30,3 +30,18 @@ configuration utilities. The actual meaning of the default (compatible to
 BRS-I) configuration would be highly specific to the vendor. In this example,
 it could capping all harts at the same performance level, or it could mean
 disabling efficiency harts.
+
+[[recipe-brs-m-guidance]]
+===== BRS-M Recipe Guidance
+
+In contrast to BRS-I and BRS-B, BRS-M has no implementation of UEFI related
+services. Non-UEFI firmware, such as *coreboot*, can also supports boot well.
+Although non-UEFI firmware does not support UEFI interface, it can also
+provide ACPI, SMBIOS and SBI services. ACPI and SMBIOS can be passed through
```

Comment:
Because this is a spec, you need to say how that would be done (or reference another spec). Again, as per my comment in https://github.com/riscv-non-isa/riscv-brs/pull/72, if you want to go this route to enable cloud-scale LinuxBoot, follow the Arm SystemReady LR appoach with their LBBR requirements and use an even more constrained subset of UEFI (compared to EBBR). You still get the consistent hand-off interface, and you don't need to hack up booting OSes to understand your new hand-off structures.

https://developer.arm.com/documentation/den0044/latest, see 10.1 LBBR-v1 requirements

Reply:
In fact, UEFI is not supported in coreboot and will not be supported in the future.

See https://lore.kernel.org/lkml/ZKgLKvBoWKSxzm6r@sunil-laptop/T/
As Coreboot maintainer Ron said: 

part1:
ACPI was not originally part of UEFI. ACPI works just fine on Chromebooks, and has for 12 years, and on coreboot since 2006, without UEFI. I've integrated support for ACPI into several code bases, including Plan 9 on non-UEFI systems.

I am unable to understand the claim that ACPI on non-UEFI RISC-V is a bad idea. Clearly, I am not alone. 

part2:
Glad to see we're not going to revisit UEFI arguments. I've been doing this dance for 25 years, and I've seen the suggestion that adding UEFI plumbing to coreboot is easier, and we've actually tried that out several times in the last 16 years, and ... it's really not worked well for many reasons, some of them involving the many ambiguities in the spec that are answered in proprietary Intel code (which Intel engineers have pointed out to me more than once).  

And that's why since 2012 chromebooks have run coreboot, without any UEFI interfaces of any kind, even though the very first chromebooks were based on UEFI. Whether or not UEFI is the standard, it is hardly universal.

I'm trying hard not to get sucked into this argument, but it is, at least, impolite, to accuse us of engaging in hand-wavery. There's probably more UEFI expertise in coreboot and oreboot than most places, and that's precisely why we want a non-UEFI option. Familiarity with UEFI did not breed respect.

So can we get back to the business at hand? bytedance may be the company proposing this change, but they are certainly not the only company interested in seeing it happen. A project I'm involved in needs this work to go through.

Reply:
On x86 Coreboot can launch U-Boot (https://u-boot.readthedocs.io/en/latest/board/coreboot/index.html). The same could be implemented for RISC-V. U-Boot can provides the UEFI API. Coreboot and UEFI are just different firmware levels and not contradictory.

For UEFI we have a standardized way to pass ACPI and SMBIOS tables to Linux. According to [RISC-V Kernel Boot Requirements and Constraints](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/riscv/boot.rst?h=v6.6-rc1) such an interface does not exist for the legacy entry point.

Referring to ACPI makes only sense if such an interface is defined or linked to.

Reply:
Because UEFI is not the only firmware interface, UEFI is not used in all application scenarios. So we want to support more firmware solutions than UEFI.

Reply:
We believe that the firmware ecosystem is developed and should not be limited to UEFI alone. In this way, it can also adapt to more scenarios.
From the BRS spec, there is no direct description of UEFI and ACPI binding. For example, coreboot can easily support ACPI/SMBIOS, but if you add EFI stub, it adds complexity and no better benefits.

Reply:
To shorten my comments: prove that you need a different way of approaching things than the Arm LBBR, which seems to tick all the checkboxes you desire (no full UEFI, suitable for Linux-based or alternative firmware), and is an industry accepted approach. You can't just propose a completely ad-hoc mechanism and not expect pushback towards something that has wider acceptance.

Reply:
Btw, would you be so kind and subscribe to tech-brs@lists.riscv.org? We could have discussed this in the meeting today and  listened to the proposal, but it wasn't brought up as an agenda item (nor did you get back to us on our request to have this presented).

My recommendation is to quench the emotions a bit, avoid hyperbole and try to understand each other. This is where, again, a presentation to the BRS team is going to be helpful. The overall approach to the BRS spec is to adopt existing functional solutions and stick to industry approaches instead of inventing new ones. This is why BRS largely copies the Arm BBR approach, and why we take an aggressive policy of not duplicating requirements seen in upstream specs (but *do* provide a non-binding guidance document to help people navigate the specs). Again, to reiterate - I personally get your reasons, but there's already an industry approach (LBBR) that I think you should investigate first before rolling your own, and have concrete feedback on what exactly doesn't work about it for you (because it sure works for other hyprescalers). I pointed you to the LBBR section, it's pretty short.

Reply:
Btw, as part of the BRS/PRS work we have identified gaps in UEFI/ACPI and have worked to address these. So, if you have concrete examples of something that is "the many ambiguities in the spec that are answered in proprietary Intel code", we'd be happy to take that flag up in the USWG (or you can raise these yourself, either via UEFI Forum USWG or the OSS Code First Process @ https://github.com/tianocore/tianocore.github.io/wiki/EDK-II-Code-First-Process).

Reply:
"Because this is a spec, you need to say how that would be done (or reference another spec). "
Generally, in a spec, you set requirements -- you need to say that it must be done -- but not how. Perhaps I am misunderstanding your wording.
In any event, the wording is "it can also provide ACPI, SMBIOS and SBI services." -- operative word, "can"

For a platform I am working on, I don't want the mess that comes with ACPI, SMBIOS, and I am trying to have SBI do as little as possible. ACPI in particular, due to ACPI ML, is a serious problem to be dealt with by removal.

Further, the coreboot port, which had an SBI that predates OpenSBI, showed how it could be done: it conformed to the interface defined in RISC-V standards. 

I'd be just as happy with wording to this effect: "platforms must support minimal SBI standards, as defined in the RISC-V privelege specification; ACPI and SMBIOS are optional." or some such and leave it that. Will that do?

Reply:
Ron, if you have a platform where you don't want ACPI and SMBIOS (i.e. you are using DT), then this use case fits squarely into the BRS-B (EBBR) profile. For BRS-B, ACPI and SMBIOS are optional. It's BRS-I where ACPI and SMBIOS are mandatory.

Reply:
> We believe that the firmware ecosystem is developed and should not be limited to UEFI alone. In this way, it can also adapt to more scenarios. From the BRS spec, there is no direct description of UEFI and ACPI binding. For example, coreboot can easily support ACPI/SMBIOS, but if you add EFI stub, it adds complexity and no better benefits.

FWIW, adding an alternate ACPI discovery mechanism itself adds complexity and no real benefit. It just shifts the problem elsewhere - namely to OSVs or cloud service providers. With an LBBR-like approach, you can still boot an unmodified Ubuntu or w/e installer/installation, with the stock kernel, etc. With your approach, you can't even boot anything unless you roll a custom kernel, or if your proposed changes are accepted and then if picked up by a distro (which is a lot of 'ifs').

A few weeks ago I asked the technical cost of implementing a minimal LBBR-like EFI stub in your firmware. I haven't seen any responses to that, but I do have personal experience doing so in an AArch64 U-boot fork about 10 years ago (e.g. EFI stub with dummy RT, used just to discover ACPI tables - this predates U-Boot's much more complete EFI implementation). The cost is insignificant - on the order of a half a day of work and 100 lines of code, if that. And the value is clear, especially when you avoid needless changes in the OSes you'd be booting that way. I can't imagine any kernel maintainer for a Linux distro agreeing that /another/ mechanism of kernel/firmware handoff is a good use of anyone's time. Feel free to loop those people in here from Canonical, Suse, Red Hat, Debian, NetBSD, etc, if I'm incorrect.

The purpose behind the BRS standard exist to make it easier to bring existing OSes and OS vendors to the nascent RISC-V ecosystem - it's a design choice to adopt known-good solutions instead of hand-rolling things unless there's really no alternative, to minimize tech debt incurred to support new systems and use-cases.

So far the only action item I see here is to continue to explore the CSP/cloud/LinuxBoot use-case. We should take LBBR as a start, again because it's already an accepted approach, and also because it doesn't suffer from NIH. Given our hopes around ratification schedule for BRS, I'd recommend getting a draft of something reviewable soon, else there's a good chance it won't make it into the first version of BRS.